bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -2,454 +2,170 @@
-			Transact {
-				origin_kind: OriginKind::Xcm,
-				call: (
-					create_call_index,
-					asset_id,
-					MultiAddress::<[u8; 32], ()>::Id(owner),
-					MINIMUM_DEPOSIT,
-				)
-					.encode()
-					.into(),
-			},
-			// Forward message id to Asset Hub
-			SetTopic(message_id.into()),
-			// Once the program ends here, appendix program will run, which will deposit any
-			// leftover fee to snowbridge sovereign.
-		]
-		.into();
+	fn convert(message: Message) -> Result<Xcm<()>, ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| ConvertMessageError::InvalidVersionedXCM)?;
+			message_xcm = versioned_xcm.try_into().map_err(|_| ConvertMessageError::InvalidXCM)?;
+		}
 
-		(xcm, total_amount.into())
-	}
+		log::debug!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
 
-	fn convert_send_token(
-		message_id: H256,
-		chain_id: u64,
-		token: H160,
-		destination: Destination,
-		amount: u128,
-		asset_hub_fee: u128,
-	) -> (Xcm<()>, Balance) {
-		let network = Ethereum { chain_id };
-		let asset_hub_fee_asset: Asset = (Location::parent(), asset_hub_fee).into();
-		let asset: Asset = (Self::convert_token_address(network, token), amount).into();
+		let network = EthereumNetwork::get();
 
-		let (dest_para_id, beneficiary, dest_para_fee) = match destination {
-			// Final destination is a 32-byte account on AssetHub
-			Destination::AccountId32 { id } =>
-				(None, Location::new(0, [AccountId32 { network: None, id }]), 0),
-			// Final destination is a 32-byte account on a sibling of AssetHub
-			Destination::ForeignAccountId32 { para_id, id, fee } => (
-				Some(para_id),
-				Location::new(0, [AccountId32 { network: None, id }]),
-				// Total fee needs to cover execution on AssetHub and Sibling
-				fee,
-			),
-			// Final destination is a 20-byte account on a sibling of AssetHub
-			Destination::ForeignAccountId20 { para_id, id, fee } => (
-				Some(para_id),
-				Location::new(0, [AccountKey20 { network: None, key: id }]),
-				// Total fee needs to cover execution on AssetHub and Sibling
-				fee,
-			),
-		};
+		let origin_location = Location::new(2, GlobalConsensus(network))
+			.push_interior(AccountKey20 { key: message.origin.into(), network: None })
+			.map_err(|_| ConvertMessageError::InvalidXCM)?;
 
-		let total_fees = asset_hub_fee.saturating_add(dest_para_fee);
-		let total_fee_asset: Asset = (Location::parent(), total_fees).into();
-		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
+		let network = EthereumNetwork::get();
 
+		let fee_asset = Location::new(1, Here);
+		let fee_value = 1_000_000_000u128; // TODO needs to be dry-run to get the fee but also
+									 // need to add a fee here for the dry run... Chicken/egg problem?
+		let fee: Asset = (fee_asset, fee_value).into();
```

Comment:
IIUC the `dry_run` on BH is to convert the InboundMessage to a XCM.  Then UI can dry run that XCM on AH for the cost.
 
So IMO `InboundMessage` should contain the fee like V1 before,  it can be zero for estimation and filled in with the result for calling the Gateway contract.

Reply:
@yrong do you think the AH execution fee (which can be 0 for estimation) should be added as a field in [MessageV2](https://github.com/claravanstaden/polkadot-sdk/pull/3/files#diff-011a2c280539a541db601aa2fd4e5f30960525c4bb35588c3de78effb01720fcR29)?

Reply:
@vgeddes or how did you imagine the AH execution fee be specified? I see there is a fee added for register token (which is the foreign asset creation deposit) but no other place in the contracts to specify the AH execution fee. Iiuc this will be the flow:

- Inbound queue runtime API to convert command to XCM
- UX takes XCM and dry-runs the XCM on AH to get fee
- UX takes fee number and adds it to the command, emits OutboundEvent

I don't see such a fee parameter in the contract code, so just wanna get your thoughts on this.

Reply:
Another option as mentioned in https://github.com/claravanstaden/polkadot-sdk/pull/3#discussion_r1851214454